Addons/tsgExoticQuadrature.hpp

```diff
@@ -178,28 +197,77 @@ inline TasGrid::CustomTabulated getExoticQuadrature(const int n, const double sh
                                     std::move(weights_cache), description);
 }
 
-// Basic function overloads.
-
-// If only the weight function is provided, the reference points and weights are generated by the GL rule and the weight_fn values
-// are generated by applying the weight function to these reference points.
+/*
+ * \ingroup TasmanianAddons
+ * \brief Constructs a CustomTabulated object containing the exotic quadrature nodes and quadrature weights for a given level \b n,
+ * scalar \b shift, a reference TasmanianSparseGrid \b grid containing loaded values of the weight function, a string \b description,
+ * and an optional bool \b is_symmetric that should be set to true if the weight function is symmetric.
+ *
+ * It is assumed that grid has the following properties:
+ *     (i)   grid.getPoints() returns the quadrature reference points used to compute the entries of the Jacobi matrix.
+ *     (ii)  grid.getQuadratureWeights() returns the quadrature reference weights corresponding to grid.getPoints().
+ *     (iii) grid.getLoadedValues() returns the values of the weight function at grid.getPoints().
+ */
+inline TasGrid::CustomTabulated getExoticQuadrature(const int n, const double shift, TasGrid::TasmanianSparseGrid grid,
+                                                    const char* description, const bool is_symmetric = false) {
+    std::vector<double> ref_points = grid.getPoints();
+    std::vector<double> shifted_weights = grid.getQuadratureWeights();
+    std::vector<double> weight_fn_vals = std::vector<double>(grid.getLoadedValues(), grid.getLoadedValues() + grid.getNumLoaded());
+    if (ref_points.size() == 0) {
+        throw std::invalid_argument("ERROR: grid needs to contain a nonzero number of nodes returned by grid.getPoints()!");
+    }
+    if (shifted_weights.size() != ref_points.size()) {
```

Comment:
* start with the error checking and raise the appropriate exception before using `grid`, you may get an error when initializing the `ref_points` or `shifted_weights`.
* there's no need to check if the shifted weights and ref_points have the same number, the TasmanianSparseGrid class always returns the same number of points and weights (it's tested elsewhere).
* you need to test the folling
    * `grid.getNumLoaded() > 0` which means there are points loaded into the grid
    * `grid.getNumDimensions() == 1` since the weights must be single dimensional
    * `grid.getNumOutputs() == 1` since the output must be a scalar
    * positivity of the weights after the shift

Reply:
Changed.